pyvrp/cpp/DurationSegment.h

```diff
@@ -132,25 +137,48 @@ DurationSegment DurationSegment::merge(Matrix<Duration> const &durationMatrix,
 
     // edgeDuration is the travel duration from our last to the other's first
-    // client, and atOther the time (after starting from our first client) at
-    // which we arrive there.
+    // client
     Dur const edgeDuration = durationMatrix(idxLast_, other.idxFirst_);
-    Dur const atOther = duration_ - timeWarp_ + edgeDuration;
 
-    // Time warp increases if we arrive at the other's first client after its
-    // time window closes, whereas wait duration increases if we arrive there
-    // before opening.
-    Dur const diffTw = std::max<Dur>(twEarly_ + atOther - other.twLate_, 0);
-    Dur const diffWait = other.twEarly_ - atOther > twLate_
-                             ? other.twEarly_ - atOther - twLate_
-                             : 0;  // ternary rather than max avoids underflow
+    // We must wait if we arrive before the earliest start time of the other
+    // segment. Note that the earliest start time of the other segment may not
+    // be the opening of the time window of the first client, but rather the
+    // time we should start to avoid unnecessary waiting time. We may still
+    // start earlier at that other segment, but this suggests that we should
+    // wait at another moment for the same total duration since we cannot
+    // finish the other route segment earlier.
+    Dur const waitDuration
+        = latestFinish_ < other.earliestStart_ - edgeDuration
+              ? other.earliestStart_ - latestFinish_ - edgeDuration
+              : 0;  // ternary rather than max avoids underflow
+
+    // The new earliest start time for the merged segment is either the
+    // earliest start time for the first/current segment or a later start time
+    // computed backwards from the earliest start time of the second/other
+    // segment. In this computation, we can ignore existing time warp in the
+    // first segment, since if there is existing time warp, there is no slack
+    // in the first segment and the earliest start time will not change.
+    Dur const mergedEarliestStart = std::max(other.earliestStart_ - waitDuration
+                                                 - edgeDuration - duration_,
+                                             earliestStart_);
+
+    // The new latest finish time for the merged segment is either the latest
+    // finish time for the second/other segment or an earlier finish time
+    // computed forward from the latest finish time of the first/current
+    // segment. In this computation, we can ignore existing time warp in the
+    // second segment, since if there is existing time warp, there is no slack
+    // in the second segment and the latest finish time will not change.
+    Dur const diffDuration = edgeDuration + waitDuration + other.duration_;
+    Dur const mergedLatestFinish
+        = latestFinish_ < other.latestFinish_ - diffDuration
+              ? latestFinish_ + diffDuration
+              : other.latestFinish_;  // Avoid overflow
```

Comment:
This is
```cpp
std::max(latestFinish_ + diffDuration, other.latestFinish_);
```
Can `latestFinish_ + diffDuration` overflow?

Reply:
Yes, `latestFinish_` can be infinite if for the depot there is no time window (correct me if wrong). Then adding a positive duration would overflow.